### ISSUE TITLE
Are complex variables supported in eager mode?

### ISSUE THREAD

**Issue by ziofil** [tensorflow/tensorflow#32774]
**System information**
- TensorFlow version (you are using): 2.0.0-rc1
- Are you willing to contribute it (Yes/No): No

**Describe the feature and the current behavior/state.**

Currently (unless I'm not doing things correctly) it seems that there is no support for complex variables. This fails:
```python
    a = tf.Variable(4.0+1.0j) # a = tf.Variable(4.0) works

    def f(x):
      return x*a

    loss = lambda: tf.abs(f(5)-5)

    optimizer = tf.optimizers.Adam(learning_rate=0.1)
    optimizer.minimize(loss, [a])
```

> WARNING:tensorflow:The dtype of the watched tensor must be floating (e.g. tf.float32), got tf.complex128

and 
> ValueError: Invalid type tf.complex128 for Variable:0, expected: [tf.float32, tf.float64, tf.float16, tf.bfloat16].

**Will this change the current api? How?**
Potentially yes

**Who will benefit with this feature?**
All of the users who used to rely on complex numbers!

**Comment by jvishnuvardhan** [comment]
I was able to reproduce the issue in `2.0.0rc2`. Please check the [gist here](https://colab.sandbox.google.com/gist/jvishnuvardhan/ecf15b48633de8064ea99a5834c523cc/32774.ipynb). Thanks!

Here is the error trace
```
WARNING:tensorflow:The dtype of the source tensor must be floating (e.g. tf.float32) when calling GradientTape.gradient, got tf.complex128
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-23-f65531aefcec> in <module>()
      5 loss = lambda: tf.abs(f(5)-5)
      6 optimizer = tf.optimizers.Adam(learning_rate=0.1)
----> 7 optimizer.minimize(loss, [a])

2 frames
/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/keras/optimizer_v2/optimizer_v2.py in _assert_valid_dtypes(self, tensors)
    830       if dtype not in valid_dtypes:
    831         raise ValueError("Invalid type %r for %s, expected: %s." %
--> 832                          (dtype, t.name, [v for v in valid_dtypes]))
    833 
    834   def _valid_dtypes(self):

ValueError: Invalid type tf.complex128 for Variable:0, expected: [tf.float32, tf.float64, tf.float16, tf.bfloat16].
```

**Comment by fchollet** [comment]
Unassigning, as I have no context on complex variable support. Sorry.

**Comment by ziofil** [comment]
@fchollet what do you mean by context?

**Comment by ziofil** [comment]
might be related https://github.com/tensorflow/tensorflow/pull/30107

**Comment by jaingaurav** [comment]
@ziofil: That issue is slightly unrelated. I have a fix that add support for complex values in (some) optimizers that should be landing shortly.

**Comment by Davidvdrm** [comment]
@jaingaurav Is there a timeframe on this fix? I am currently looking at a fix myself and wonder if it is worth it.

**Comment by jaingaurav** [comment]
@Davidvdrm: The fix is already committed and available in the nightly pip package. It will also be included in the upcoming 2.1 release.

**Comment by Davidvdrm** [comment]
I built the nightly package and still get a NotFoundError concerning the needed ApplyGradient ops for Adam, SGD and AdaDelta. This is the same issue as before the upgrade to the nightly build and I created a new issue [https://github.com/tensorflow/tensorflow/issues/34098](https://github.com/tensorflow/tensorflow/issues/34098) concerning this problem. Should the change fix this issue or are they unrelated?

Full error code:
```
> NotFoundError                             Traceback (most recent call last)
> <ipython-input-41-0c0c662a0b5d> in <module>
>      26         # Optimize the model
>      27         loss_value, grads = grad(model, x, y)
> ---> 28         optimizer.apply_gradients(zip(grads, model.trainable_variables))
>      29 
>      30         # Track progress
> 
> ~\AppData\Roaming\Python\Python37\site-packages\tensorflow_core\python\keras\optimizer_v2\optimizer_v2.py in apply_gradients(self, grads_and_vars, name)
>     442           functools.partial(self._distributed_apply, apply_state=apply_state),
>     443           args=(grads_and_vars,),
> --> 444           kwargs={"name": name})
>     445 
>     446   def _distributed_apply(self, distribution, grads_and_vars, name, apply_state):
> 
> ~\AppData\Roaming\Python\Python37\site-packages\tensorflow_core\python\distribute\distribute_lib.py in merge_call(self, merge_fn, args, kwargs)
>    1947     if kwargs is None:
>    1948       kwargs = {}
> -> 1949     return self._merge_call(merge_fn, args, kwargs)
>    1950 
>    1951   def _merge_call(self, merge_fn, args, kwargs):
> 
> ~\AppData\Roaming\Python\Python37\site-packages\tensorflow_core\python\distribute\distribute_lib.py in _merge_call(self, merge_fn, args, kwargs)
>    1954         distribution_strategy_context._CrossReplicaThreadMode(self._strategy))  # pylint: disable=protected-access
>    1955     try:
> -> 1956       return merge_fn(self._strategy, *args, **kwargs)
>    1957     finally:
>    1958       _pop_per_thread_mode()
> 
> ~\AppData\Roaming\Python\Python37\site-packages\tensorflow_core\python\keras\optimizer_v2\optimizer_v2.py in _distributed_apply(self, distribution, grads_and_vars, name, apply_state)
>     486           update_ops.extend(
>     487               distribution.extended.update(
> --> 488                   var, apply_grad_to_update_var, args=(grad,), group=False))
>     489 
>     490       any_symbolic = any(isinstance(i, ops.Operation) or
> 
> ~\AppData\Roaming\Python\Python37\site-packages\tensorflow_core\python\distribute\distribute_lib.py in update(self, var, fn, args, kwargs, group)
>    1541       kwargs = {}
>    1542     with self._container_strategy().scope():
> -> 1543       return self._update(var, fn, args, kwargs, group)
>    1544 
>    1545   def _update(self, var, fn, args, kwargs, group):
> 
> ~\AppData\Roaming\Python\Python37\site-packages\tensorflow_core\python\distribute\distribute_lib.py in _update(self, var, fn, args, kwargs, group)
>    2172     # The implementations of _update() and _update_non_slot() are identical
>    2173     # except _update() passes `var` as the first argument to `fn()`.
> -> 2174     return self._update_non_slot(var, fn, (var,) + tuple(args), kwargs, group)
>    2175 
>    2176   def _update_non_slot(self, colocate_with, fn, args, kwargs, should_group):
> 
> ~\AppData\Roaming\Python\Python37\site-packages\tensorflow_core\python\distribute\distribute_lib.py in _update_non_slot(self, colocate_with, fn, args, kwargs, should_group)
>    2178     # once that value is used for something.
>    2179     with UpdateContext(colocate_with):
> -> 2180       result = fn(*args, **kwargs)
>    2181       if should_group:
>    2182         return result
> 
> ~\AppData\Roaming\Python\Python37\site-packages\tensorflow_core\python\keras\optimizer_v2\optimizer_v2.py in apply_grad_to_update_var(var, grad)
>     468       if "apply_state" in self._dense_apply_args:
>     469         apply_kwargs["apply_state"] = apply_state
> --> 470       update_op = self._resource_apply_dense(grad, var, **apply_kwargs)
>     471       if var.constraint is not None:
>     472         with ops.control_dependencies([update_op]):
> 
> ~\AppData\Roaming\Python\Python37\site-packages\tensorflow_core\python\keras\optimizer_v2\adadelta.py in _resource_apply_dense(self, grad, var, apply_state)
>     139         coefficients['epsilon'],
>     140         grad,
> --> 141         use_locking=self._use_locking)
>     142 
>     143   def _resource_apply_sparse(self, grad, var, indices, apply_state=None):
> 
> ~\AppData\Roaming\Python\Python37\site-packages\tensorflow_core\python\training\gen_training_ops.py in resource_apply_adadelta(var, accum, accum_update, lr, rho, epsilon, grad, use_locking, name)
>    1100         pass  # Add nodes to the TensorFlow graph.
>    1101     except _core._NotOkStatusException as e:
> -> 1102       _ops.raise_from_not_ok_status(e, name)
>    1103   # Add nodes to the TensorFlow graph.
>    1104   if use_locking is None:
> 
> ~\AppData\Roaming\Python\Python37\site-packages\tensorflow_core\python\framework\ops.py in raise_from_not_ok_status(e, name)
>    6596   message = e.message + (" name: " + name if name is not None else "")
>    6597   # pylint: disable=protected-access
> -> 6598   six.raise_from(core._status_to_exception(e.code, message), None)
>    6599   # pylint: enable=protected-access
>    6600 
> 
> ~\Anaconda3\lib\site-packages\six.py in raise_from(value, from_value)
> 
> 
> NotFoundError: No registered 'ResourceApplyAdadelta' OpKernel for 'CPU' devices compatible with node {{node ResourceApplyAdadelta}}
> 	 (OpKernel was found, but attributes didn't match) Requested Attributes: T=DT_COMPLEX64, use_locking=true
> 	.  Registered:  device='CPU'; T in [DT_HALF]
>   device='CPU'; T in [DT_BFLOAT16]
>   device='CPU'; T in [DT_FLOAT]
>   device='CPU'; T in [DT_DOUBLE]
>  [Op:ResourceApplyAdadelta]
```

`
>
